### PR TITLE
[RTC-275] Allow for subscribing to notifications from a single room

### PR DESCRIPTION
**Pull request by LVala**

### CODE REVIEW

File: lib/jellyfish/notifier.ex [hunk 1]
```diff
@@ -97,7 +145,7 @@ defmodule Jellyfish.Notifier do
            apply(WebSockex, fun, ["#{address}/socket/server/websocket", __MODULE__, state]),
          :ok <- WebSockex.send_frame(pid, {:binary, auth_msg}) do
       receive do
-        {:jellyfish, %Authenticated{}} ->
```
Comment (roznawsk) [inline lib/jellyfish/notifier.ex:100]: Why we no longer use the `Authenticated` message?

File: test/jellyfish/notifier_test.exs [hunk 2]
```diff
@@ -21,17 +21,88 @@ defmodule Jellyfish.NotifierTest do
     defstruct [:qwe, :rty]
   end
 
-  setup do
-    {:ok, notifier_pid} = Jellyfish.Notifier.start()
+  describe "connecting to the server" do
+    setup do
+      address = Application.fetch_env!(:jellyfish_server_sdk, :server_address)
+      token = Application.fetch_env!(:jellyfish_server_sdk, :server_api_token)
```
Comment (roznawsk) [inline test/jellyfish/notifier_test.exs:27]: I think we can just start notifier without passing any options, it will fetch them from the env anyway.

File: test/jellyfish/notifier_test.exs [hunk 3]
```diff
@@ -21,17 +21,88 @@ defmodule Jellyfish.NotifierTest do
...
+      assert {:error, :invalid_token} =
+               Notifier.start(server_address: address, server_api_token: "invalid_token")
+    end
   end
 
-  describe "peer notifications" do
-    test "peer connects and then disconnects", %{client: client} do
+  describe "subscribing" do
+    setup do
+      {:ok, notifier} = Notifier.start_link()
```
Comment (roznawsk) [inline test/jellyfish/notifier_test.exs:45]: IMO you can follow this way of starting notifier (i.e. without passing options) in other places.

File: test/jellyfish/notifier_test.exs [hunk 4]
```diff
@@ -21,17 +21,88 @@ defmodule Jellyfish.NotifierTest do
     defstruct [:qwe, :rty]
```
Comment (roznawsk) [inline test/jellyfish/notifier_test.exs:21]: Is this struct used anywhere?

**Review by roznawsk — CHANGES_REQUESTED** [review]
I left some nitpicks.
I am wondering if it would be useful to test other notifications in the tests, other than `PeerConnected`. Can we assume that since this one notification is sent and handled properly, all others will be as well?

File: lib/jellyfish/notifier.ex (hunk 1 above)
Comment (LVala) [reply lib/jellyfish/notifier.ex:100]: This message was only received by the process that started the `Notifier` process - a hack to make the `start/1` and `start_link/1` blocking. Nothing changes in the API when it comes to the messages, it wasn't available to the user before either.

File: test/jellyfish/notifier_test.exs (hunk 4 above)
Comment (LVala) [reply test/jellyfish/notifier_test.exs]: It seems like it isn't.

**Comment by LVala** [comment]
> I left some nitpicks. I am wondering if it would be useful to test other notifications in the tests, other than `PeerConnected`. Can we assume that since this one notification is sent and handled properly, all others will be as well?

Well, we could, but all of the messages (except for `%Authenticated{}`) are handled by the same function and simply decoded and send further. I guess that if somebody messed with this function and broke it, it could go unnoticed. However, I don't thing we can trigger the other notifications easily (it's `(Room|Component|Peer)Crashed`).

**Review by roznawsk — APPROVED** [review]
Looks great! 🎖️

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: lib/jellyfish/notifier.ex [hunk 5]
```diff
@@ -60,34 +57,85 @@ defmodule Jellyfish.Notifier do
...
+  if this parameter is passed, notifications will be filtered only to this specific room.
+  """
+  @type subscribe_options :: [process: Process.dest(), room_id: Room.id()]
+
+  @doc """
+  Subscribes the process to receive server notifications.
+
+  Notifications are sent to the process in a form of `{:jellyfish, msg}`,
+  where `msg` is one of the structs defined in `lib/proto/jellyfish/server_notifications.pb.ex`,
+  for example `{:jellyfish, %Jellyfish.ServerMessage.RoomCrashed{room_id: "some_id"}}`.
```
[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/notifier.ex:74]: That's exactly what I initially did, although the types of generated structs contain references to types from the `protobuf` library and breaks CI etc. But I'll try to find a workaround.

File: lib/jellyfish/notifier.ex [hunk 6]
```diff
@@ -60,34 +57,85 @@ defmodule Jellyfish.Notifier do
...
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_cast({:subscribe, pid, room_id}, state) do
+    state =
+      state.subscriptions
+      |> Map.get(room_id, MapSet.new())
+      |> MapSet.put(pid)
+      |> then(&put_in(state, [:subscriptions, room_id], &1))
```
Comment (bblaszkow06) [inline lib/jellyfish/notifier.ex:110]: I'd replace that pipe with `update_in`:
```suggestion
      update_in(state.subscriptions[room_id], fn
        nil -> MapSet.new()
        set -> MapSet.put(set, pid)
      end)
```

File: lib/jellyfish/notifier.ex [hunk 7]
```diff
@@ -60,34 +57,85 @@ defmodule Jellyfish.Notifier do
...
   end
 
   @impl true
-  def handle_cast(_msg, state) do
-    # ignore incoming messages
+  def handle_cast({:unsubscribe, pid}, state) do
+    state =
+      state.subscriptions
+      |> Map.new(fn {id, pids} -> {id, MapSet.delete(pids, pid)} end)
+      |> then(&Map.put(state, :subscriptions, &1))
```
Comment (bblaszkow06) [inline lib/jellyfish/notifier.ex:120]:
```suggestion
      Map.update!(state, :subscriptions, fn subs ->
        Map.new(subs, fn {id, pids} -> {id, MapSet.delete(pids, pid)} end)
      end)
```

File: lib/jellyfish/notifier.ex [hunk 8]
```diff
@@ -60,34 +63,78 @@ defmodule Jellyfish.Notifier do
...
+  end
+
+  @doc """
+  Unsubscribe from receiving notifications requested by calling `subscribe/2`.
+
+  Stops the `Notifier` from sending any notifications to `process`.
+  """
+  @spec unsubscribe(notifier(), Process.dest()) :: :ok
+  def unsubscribe(notifier, process \\ self()) do
+    WebSockex.cast(notifier, {:unsubscribe, process})
```
Comment (bblaszkow06) [inline lib/jellyfish/notifier.ex:92]: There's one more RC here - the subscriber might have some notifications already queued in its mailbox or may receive it before the notifier processes this message (since it's a cast, not a call)

File: lib/jellyfish/notifier.ex [hunk 9]
```diff
@@ -60,34 +63,78 @@ defmodule Jellyfish.Notifier do
...
   def handle_frame({:binary, msg}, state) do
     %ServerMessage{content: {_type, notification}} = ServerMessage.decode(msg)
+    handle_notification(notification, state)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_cast({:subscribe, pid, room_id}, state) do
+    state =
```
Comment (bblaszkow06) [inline lib/jellyfish/notifier.ex:105]: There's one more important thing missing - the notifier should monitor the subscriber and remove it when it goes down

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: lib/jellyfish/notification.ex [hunk 10]
```diff
@@ -0,0 +1,81 @@
+defmodule Jellyfish.Notification do
+  @moduledoc """
+  Submodules of this module represent Jellyfish server notifications.
+  See `Jellyfish.Notifier` for more information.
+  """
```
[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/notification.ex:5]: I kept it only so the reference in `doc` of `Notifier.subscribe/3` works, but I guess its not really helpful, I'll find a workaround.

File: lib/jellyfish/notification.ex [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+  defmodule RoomCrashed do
+    @moduledoc "Defines `t:Jellyfish.Notification.RoomCrashed.t/0`."
```
[the remark this answers is not in the document]
Comment (LVala) [reply lib/jellyfish/notification.ex:8]: I was initially going to leave the `moduledoc`s empty, but `credo` is not happy about that. Thb, I just discovered `@moduledoc nil` and the module names and fields are quite self-explanatory.

File: README.md [hunk 12]
```diff
@@ -30,7 +30,8 @@ and receive server notifications:
 
 ```elixir
 # start process responsible for receiving notifications
-{:ok, _pid} = Jellyfish.Notifier.start(server_address: "localhost:5002", server_api_key: "your-jellyfish-token")
+{:ok, notifier} = Jellyfish.Notifier.start(server_address: "localhost:5002", server_api_key: "your-jellyfish-token")
+:ok = Jellyfish.Notifier.subscribe(notifier, :all)
```
[the remark this answers is not in the document]
Comment (roznawsk) [reply README.md:34]: I don't think that's a problem. When subscribing to notifications we can always get the room's state to be up to date.

[the remark this answers is not in the document]
Comment (LVala) [reply README.md:34]: I mean, it's a little bit inconvenient, imagine that you created a room, it crashed, and only then you subscribed to notifications. When you query the room state, you'll get that it doesn't exist.

[the remark this answers is not in the document]
Comment (LVala) [reply README.md:34]: The only thing that's comes to my mind is buffering a few of the last messages tho, as the messages that we might lose aren't very frequent (basically only `RoomCrashed`, maybe something like `RoomRemoved` in the future), what do you think @bblaszkow06?

File: lib/jellyfish/notifier.ex (hunk 8 above)
Comment (LVala) [reply lib/jellyfish/notifier.ex:92]: `Websockex` API doesn't provide `call` for some reason, besides that, even if we used `call`, I don't see how we can take care of the messages in a mailbox, as we can call the `unsubscribe` for process other than `self()`?
But is this really a problem, I didn't thought of `unsubscribe` as a "after this point you definitely won't get notifications" but rather "tell the notifier to not spam this process anymore", especially as every notification contains `room_id` anyway.

Comment (bblaszkow06) [reply lib/jellyfish/notifier.ex:92]: > But is this really a problem, I didn't thought of unsubscribe as a "after this point you definitely won't get notifications" but rather "tell the notifier to not spam this process anymore", especially as every notification contains room_id and unwanted messages can be ignored

Well, the minimal solution would be to document that behaviour and what you thought of it.
Taking one step backwards, the subscription of process other than self seems like something rarely used and if we got rid of that, we could flush the mailbox like [`Process.demonitor/2`](https://hexdocs.pm/elixir/Process.html#demonitor/2)

API-wise it might be surprising to receive the messages you've just unsubscribed from.

Going even further back, I'm not sure the unsubscribe is needed at all. Most of the time the combination of `subscribe` and monitoring the subscriber will be enough.

To sum up, I'd probably remove the unsubscription in favour of monitoring, then if needed added the unsubscribe for the calling process with flush option and then if necessary add an option to unsub other process

File: .gitmodules [hunk 13]
```diff
@@ -1,4 +1,4 @@
 [submodule "protos"]
 	path = protos
 	url = https://github.com/jellyfish-dev/protos.git
-	branch = master
+	branch = per-room-notifications
```
Comment (LVala) [inline .gitmodules:4]: This is going to be reverted before merging, also tests doesn't pass because they require new Jellyfish image from PR which hasn't been merged yet.

File: lib/jellyfish/notifier.ex [hunk 14]
```diff
@@ -60,34 +69,82 @@ defmodule Jellyfish.Notifier do
...
 
-    send(state.receiver_pid, {:jellyfish, notification})
     {:ok, state}
   end
 
   @impl true
-  def handle_cast(_msg, state) do
-    # ignore incoming messages
+  def handle_cast({:subscribe, pid, room_id}, state) do
+    state = put_in(state.pending_subscriptions[room_id], pid)
```
Comment (LVala) [inline lib/jellyfish/notifier.ex:104]: I just noticed that this has to be fixed, multiple processes can subscribe to the same `room_id`.

File: lib/jellyfish/notifier.ex [hunk 15]
```diff
@@ -112,4 +169,73 @@ defmodule Jellyfish.Notifier do
...
+  defp from_room_state_message(msg) do
+    peers =
+      msg.peers
+      |> Enum.map(
+        &%Peer{id: &1.id, type: from_proto_type(&1.type), status: from_proto_status(&1.status)}
+      )
+
+    components =
+      msg.components
+      |> Enum.map(&%Component{id: &1.id, type: from_proto_type(&1.type)})
```
Comment (bblaszkow06) [inline lib/jellyfish/notifier.ex:225]: IMO these should be implemented in Peer and Component, i.e.

```suggestion
      |> Enum.map(&Peer.from_proto/1)

    components =
      msg.components
      |> Enum.map(&Component.from_proto/1)
```

File: lib/jellyfish/notifier.ex [hunk 16]
```diff
@@ -112,4 +169,73 @@ defmodule Jellyfish.Notifier do
...
+
+    config =
+      msg.config
+      |> Map.from_struct()
+      |> Map.reject(fn {k, _v} -> k == :__unknown_fields__ end)
```
Comment (bblaszkow06) [inline lib/jellyfish/notifier.ex:230]: I'd use drop or pop here
```suggestion
      |> Map.drop([:__unknown_fields__])
```

File: lib/jellyfish/notification.ex [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  def to_notification(%module{} = message) do
+    notification_module =
+      module
+      |> Module.split()
+      |> List.last()
+      |> then(&Module.concat(__MODULE__, &1))
+
+    message
+    |> Map.from_struct()
+    |> Enum.reject(fn {k, _v} -> k in @discarded_fields end)
```
Comment (bblaszkow06) [inline lib/jellyfish/notification.ex:77]: This round of review reminded me of `Map.drop`, it can be applied here as well
```suggestion
    |> Map.drop(@discarded_fields)
```

File: lib/jellyfish/notifier.ex [hunk 18]
```diff
@@ -98,7 +98,11 @@ defmodule Jellyfish.Notifier do
 
   @impl true
   def handle_cast({:subscribe, pid, room_id}, state) do
-    state = put_in(state.pending_subscriptions[room_id], pid)
+    state =
+      update_in(state.pending_subscriptions[room_id], fn
+        nil -> [pid]
+        pids -> [pid | pids]
```
Comment (LVala) [inline lib/jellyfish/notifier.ex:104]: I think that this simple queue is enough as we can safely assume that responses from server will be in the same order as sent `RoomStateRequest` messages thanks to WebSockets.

Comment (bblaszkow06) [reply lib/jellyfish/notifier.ex:104]: Well, this assumption is much stronger - in addition to transport assuring the order, the backend must ensure the processing order. I can imagine async processing of "request" on the backend that would break it.

For now though it will be enough to document that assumption - please add a comment with that

**Review by bblaszkow06 — APPROVED** [review]
